File: app/src/main/java/pl/laskowski/marcin/solving/sat/IndexEncoder.java
package pl.laskowski.marcin.solving.sat;

import pl.laskowski.marcin.model.Point;

/**
 * Created by Marcin Laskowski.
 */

public class IndexEncoder {

    private final int sizeX;
    private final int sizeY;
    private final int blockSize;

    public IndexEncoder(int sizeX, int sizeY) {
        this.sizeX = sizeX;
        this.sizeY = sizeY;
        this.blockSize = sizeX * sizeY;
    }

    public int encode(Point p, int value) {
        return p.getX() + sizeX * p.getY() + blockSize * --value + 1;
    }

    public Point decodePoint(int index) {
        int x = (index - 1) % sizeX;
        int y = ((index - 1) / sizeX) % sizeY;
        return new Point(x, y);
    }

    public int decodeValue(int index) {
        return --index / blockSize + 1;
    }

}

File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/solving/sat/IndexEncoder.kt
package com.github.ilikeyourhat.sudokusolver.solving.sat

import com.github.ilikeyourhat.sudokusolver.model.Point

class IndexEncoder(
    private val sizeX: Int,
    private val sizeY: Int
) {
    private val blockSize = sizeX * sizeY

    fun encode(p: Point, value: Int): Int {
        return p.x + sizeX * p.y + blockSize * (value - 1) + 1
    }

    fun decodePoint(index: Int): Point {
        val x = (index - 1) % sizeX
        val y = (index - 1) / sizeX % sizeY
        return Point(x, y)
    }

    fun decodeValue(index: Int): Int {
        return (index - 1) / blockSize + 1
    }
}
